resolver/type_syntax.cc

```diff
@@ -278,18 +278,20 @@ ParsedSig parseSigWithSelfTypeParams(core::Context ctx, const ast::Send &sigSend
                 bool validBind = false;
                 auto bind = getResultTypeWithSelfTypeParams(ctx, send->getPosArg(0), *parent, args);
                 if (core::isa_type<core::ClassType>(bind)) {
-                    auto classType = core::cast_type_nonnull<core::ClassType>(bind);
-                    sig.bind = classType.symbol;
+                    sig.bind = bind;
                     validBind = true;
                 } else if (auto appType = core::cast_type<core::AppliedType>(bind)) {
                     // When `T.proc.bind` is used with `T.class_of`, pass it
                     // through as long as it only has the AttachedClass type
                     // member.
                     if (appType->klass.data(ctx)->isSingletonClass(ctx) &&
                         appType->klass.data(ctx)->typeMembers().size() == 1) {
-                        sig.bind = appType->klass;
+                        sig.bind = bind;
                         validBind = true;
                     }
+                } else if (core::isa_type<core::SelfType>(bind) || core::isa_type<core::LambdaParam>(bind)) {
```

Comment:
I'll switch those to check if `send->fun` is specifically those two.

To remember those cases and propagate the information to `environment.cc`, are you thinking of some boolean fields in `ArgInfo` or did you have a different place to store this in mind?

Reply:
Something like [this](https://github.com/vinistock/sorbet/pull/new/allow_attached_bind)? I put it in a different branch before committing it to this one, since I wasn't sure it was the right direction.

Reply:
Another option that @Morriar brought up, which doesn't even need new booleans, is to create a new magic classes `BindToAttachedClass` and `BindToSelfType` that will let us know about the special binding in `environment.cc`

[Here is the exploration](https://github.com/sorbet/sorbet/compare/master...vinistock:allow_attached_bind?expand=1)